### ISSUE TITLE
Pipeline not triggered by webhook

### ISSUE THREAD

**Issue by tomgs** [telia-oss/github-pr-resource#183]
Hiya!

Deeply considering the move to Concourse for a while now, and I'm currently constructing a copy of our existing CI setup for one of our projects to give it a spin.

Our flow today is as follows in (using GitHub + Azure Pipelines):

1. PR is opened
2. Since PR includes at least one commit, PR triggers CI pipeline
3. If any step fails, CI pushes a status to GitHub checks
4. If pipeline completed, CI pushes a status to GitHub checks

I've managed to replicate the last 3 steps in Concourse in full, quite fast and using the command line on a local Concourse setup which is super-duper-awesome (much, much faster the re-doing CI configurations in Azure). 

Step 1, however, remains semi-unsolved. I've got two options to know that a PR has been opened:

1. **poll** - Use `every_which` in my resource declaration to poll the relevant resource every so-and-so seconds (where the resource's `resource-type` is `github-pr-resource`, based on this repo).
2. **push** - Wait for a webhook to come from GitHub whenever a new `push` or `pull_request` event have been created on the repo.

The `poll` version works fine, but has the obvious problem of waiting for the timeout to expire and of over-polling the GitHub API which may cause problems at the future.

The `push` version **does not work**, or at least not how I would expect it to work. I've looked at https://github.com/concourse/concourse/issues/3411 and it seems this is a similar issue to what I'm facing: I'd like the webhook to trigger the **pipeline**, no the **check** of the resource.

Some help would be much appreciated - I've set the webhook on the repo with the correct privileges, set the personal access token with the correct privileges as well, and can see that `POST`ing to the relevant webhook returns a 201 with the response I'd expect from concourse. 

My config (based on the example from the documentation):

```
resource_types:
- name: pull-request-base-resource
  type: docker-image
  source:
    repository: teliaoss/github-pr-resource

resources:
- name: pull-request
  type: pull-request-base-resource
  check_every: 10s
  webhook_token: <WEBHOOK_TOKEN> 
  source:
    repository: <MY_REPO>
    access_token: <MY_PERSONAL_ACCESS_TOKEN>
    base_branch: <MY_BRANCH>
jobs:
- name: test
  plan:
  - get: pull-request
    trigger: true
    version: latest # Tried with every as well...
  - put: pull-request
    params:
      path: pull-request
      status: pending
  - task: unit-test
    config:
      platform: linux
      image_resource:
        type: docker-image
        source: {repository: alpine/git, tag: "latest"}
      inputs:
        - name: pull-request
      run:
        path: /bin/sh
        args:
          - -xce
          - |
            cd pull-request
            git log --graph --all --color --pretty=format:"%x1b[31m%h%x09%x1b[32m%d%x1b[0m%x20%s" > log.txt
            cat log.txt
    on_failure:
      put: pull-request
      params:
        path: pull-request
        status: failure
  - put: pull-request
    params:
      path: pull-request
      status: success
```

**Comment by pivotal-jamie-klassen** [comment]
@tomgs can you describe in a bit more detail what you mean by
> The push version **does not work**, or at least not how I would expect it to work

? I'm assuming you're doing some kind of end-to-end manual test ("opening a PR and waiting for something to happen in Concourse") - what are the steps you are taking, what is the behaviour you would expect and what is the behaviour that you observe? Also it's awesome that you shared your pipeline; can you share the specific configuration you have on the GitHub side (just in case there might be a problem there - I'm sure the authors of this repo could advise)?

More specifically, can you validate that the webhook is actually being triggered? One way I could imagine to test, this based on the code [here](https://github.com/concourse/concourse/blob/master/atc/api/resourceserver/check_webhook.go), is: you could try configuring the webhook on the github side with an invalid token - then you could create a PR and look in the logs for `check-resource-webhook` expecting to see `invalid-token` somewhere.

I'm asking this because I noticed this section in the README:
> **Note on webhooks:**
This resource does not implement any caching, so it should work well with webhooks (should be subscribed to `push` and `pull_request` events).
One thing to keep in mind however, is that _pull requests that are opened from a fork and commits to said fork will not generate notifications over the webhook_. So if you have a repository with little traffic and expect pull requests from forks, you'll need to discover those versions with `check_every: 1m` for instance. `check` in this resource is not a costly operation,
 so normally you should not have to worry about the rate limit.

Is it possible that, in the scenarios you are testing, your PR is coming from a fork?

**Comment by tomgs** [comment]
Hey @pivotal-jamie-klassen, thanks for the swift response here and in the other issue!
I'll follow up here as well as on the other thread.

(Have I mentioned that Concourse is just freaking awesome already? Really hoping we can massage it to fix our use case, as it would make my day to day much more pleasant)

To answer your questions one by one:

> I'm assuming you're doing some kind of end-to-end manual test ("opening a PR and waiting for something to happen in Concourse") - what are the steps you are taking, what is the behaviour you would expect and what is the behaviour that you observe? 

Should have been clearer. Let's break this down:
**What I want:** A PR is opened targeting `master` in my repo / a commit is made to such a PR --> CI picks up on the PR being opened  _immediately_ --> Pipeline is started _immediately_
**What I have:** A PR is opened targeting `master` on my repo / a commit is made to such a PR --> CI picks up on the PR being opened _about 10 seconds after it has opened_ --> Pipeline is started _10-15 seconds or so after PR was opened_

Besides making sense from a CI architecture perspective (if A happens than immediately B happens), the scenario I want buys us precious worker time back. It also eliminates a 10-20 second window where GitHub knows a PR has been opened, but CI has not caught up with the `put` step to change the check status to pending. In that window, a developer can "bypass" CI and just click merge (since no `required` CI check is blocking the merge from happening).

The steps I'm taking to check this:

1. Create a commit on a local copy of the PR's branch
2. Pushing the commit to GitHub
3. Observing Concourse + the GitHub PR checks page, noting the aforementioned behavior

If it matters, I'm spinning up Concourse locally using the (lovely) [StarkAndWayne](https://concoursetutorial.com/) `docker-compose` setup fronm https://github.com/starkandwayne/concourse-tutorial/blob/master/docker-compose.yml. It's a bare-bones, two container setup that I use for evaluating the platform.

> Also it's awesome that you shared your pipeline; can you share the specific configuration you have on the GitHub side (just in case there might be a problem there - I'm sure the authors of this repo could advise)?

Yeah, of course - while I can't share the actual repo for obvious reasons, the general idea is this:

1. Since I'm running locally, spin up an [`ngrok`](https://ngrok.com/) instance to tunnel into my localhost's port 8080 (where concourse lives).
2. In the repo I'm testing with, go to Settings --> Webhooks --> Add webhook
3. Following Concourse's [resource webhook token docs](https://concourse-ci.org/resources.html#resource-webhook-token), I've created a webhook URL of the following format:

```
https://MY_LOCAL_NGROK_ID.ngrok.io/api/v1/teams/TEAM_NAME/pipelines/PIPELINE_NAME/resources/RESOURCE_NAME/check/webhook?webhook_token=WEBHOOK_TOKEN
```
4. I entered said URL in the webhook configuration page:
![image](https://user-images.githubusercontent.com/8835035/71672457-ab392680-2d7e-11ea-85ed-78c955a83d25.png)

Note that the `WEBHOOK_TOKEN` appended to the URL should be the same as the `Secret` I'm entering - made sure they match, and are a simple enough, short string so Concourse won't get confused (probably not needed, but just in case).
5. Created the webhook

> More specifically, can you validate that the webhook is actually being triggered? One way I could imagine to test, this based on the code here, is: you could try configuring the webhook on the github side with an invalid token - then you could create a PR and look in the logs for check-resource-webhook expecting to see invalid-token somewhere.

Ah, that's an interesting point. I figured it was enough to receive a `2**` response from Concourse to ensure the webhook's actually pinged. So what I did was:

1. `curl -X POST https://MY_LOCAL_NGROK_ID.ngrok.io/api/v1/teams/TEAM_NAME/pipelines/PIPELINE_NAME/resources/RESOURCE_NAME/check/webhook?webhook_token=WEBHOOK_TOKEN`
2. Get back nothing from `curl`
3. Observe that `ngrok` logs the request with a `2**` response, which it does - I'm getting back a `201 created`.

But let's dive deeper for a second. First, `fly -t <MY_TARGET> check-resource  -r <MY_PIPELINE>/pull-request`, seems to succeed:
```
id  name                        status     check_error                                                   │66125e75a5a1        concourse/concourse:5.6.0   "dumb-init /usr/loca…"   About a minute ago   Up 53 secon
4   pull-request-base-resource  succeeded                                                                │ds       0.0.0.0:8080->8080/tcp   spin-concourse_concourse_1                                            
id  name          status     check_error                                                                 │7401a1d3d471        postgres                    "docker-entrypoint.s…"   About a minute ago   Up 53 secon
5   pull-request  succeeded                               
``` 
`grep`ping the Concourse main container (not the db) logs for `check-resource-webhook` yields nothing. Same for `webhook`. This is bizarre, and I assume means the `pull-request` resource is not working as I expect it to. Could this be the source of the problem? I'm assuming that's out of your scope and down to the `pull-request` resource maintainers, I guess.

> Is it possible that, in the scenarios you are testing, your PR is coming from a fork?

Nope, no forks for me, strictly PRs to the main repo.

Also taking a look at the other thread you mentioned. And thanks a lot, again!

**Comment by itsdalmo** [comment]
Hi and thanks for the issue. With reference to the Concourse documentation:
https://concourse-ci.org/resources.html#resource-webhook-token

> Note that the WEBHOOK_TOKEN appended to the URL should be the same as the Secret I'm entering in the window above. I made sure the two match to the same string, and that the string is a simple and short string so Concourse won't get confused (probably not needed, but just in case).

Just FYI; you don't need to set the `Secret` in Github. Concourse only cares about matching the `WEBHOOK_TOKEN` set in your resource definition with the path variable used when delivering the webhook (from Github).

> I'd like the webhook to trigger the pipeline, no the check of the resource.

This is not possible. Concourse expects all webhooks to be delivered with a reference to the resource it should trigger, which is why the `webhook_token` is specified per resource and not per pipeline. 

> CI picks up on the PR being opened / the commit being made immediately --> Pipeline is started immediately

This is not possible because (as the docs referenced above states) when Concourse receives an incoming webhook, the payload is discarded and a regular `check` is scheduled instead, which takes time and makes Concourse less responsive; please see https://github.com/concourse/concourse/issues/2240 for more details.

Probably not the answers you were looking for, but I hope it helps anyhow 😅

**Comment by tomgs** [comment]
@itsdalmo Thank you! This makes more sense now. I'll follow the other issue, and try to figure out another way forward without this.

**Comment by afonsoma9** [comment]
I am trying to configure the concourse to execute a pipeline at each pull request and although the webhook is being sent and answered successfully the pipeline does not start.

### Request
![Captura de tela de 2020-01-17 16-36-16](https://user-images.githubusercontent.com/38052959/72640897-9a281200-3947-11ea-9050-67f9c343e705.png)
### Response
![Captura de tela de 2020-01-17 16-36-32](https://user-images.githubusercontent.com/38052959/72640914-a2804d00-3947-11ea-8da5-e1d61515aa32.png)
### My Compose Configuration
```yml
version: "3"

services:
  gitea-app:
    image: gitea/gitea:latest
    container_name: gitea-app
    environment:
      - USER_UID=1000
      - USER_GID=1000
      - ROOT_URL=http://localhost:3000
      - SSH_DOMAIN=mydomain.com
    restart: always
    volumes:
      - ./volumes/gitea_app:/data
    ports:
      - "3000:3000"
      - "22:22"
    networks:
      - appnet

  gitea-db:
    image: postgres:alpine
    container_name: gitea-db
    ports:
      - 5440:5432
    restart: always
    volumes:
      - ./volumes/gitea_db:/var/lib/postgresql/data
    environment:
      - POSTGRES_USER=postgres
      - POSTGRES_PASSWORD=postgres
      - POSTGRES_DB=gitea
    networks:
      - appnet

  concourse-db:
    image: postgres
    environment:
      POSTGRES_DB: concourse
      POSTGRES_PASSWORD: concourse_pass
      POSTGRES_USER: concourse_user
      PGDATA: /database
    networks:
      - appnet

  concourse:
    image: concourse/concourse
    command: quickstart
    privileged: true
    depends_on: [concourse-db]
    ports: ["8080:8080"]
    environment:
      CONCOURSE_POSTGRES_HOST: concourse-db
      CONCOURSE_POSTGRES_USER: concourse_user
      CONCOURSE_POSTGRES_PASSWORD: concourse_pass
      CONCOURSE_POSTGRES_DATABASE: concourse
      CONCOURSE_EXTERNAL_URL: http://localhost:8080
      CONCOURSE_ADD_LOCAL_USER: admin:admin
      CONCOURSE_MAIN_TEAM_LOCAL_USER: admin
      CONCOURSE_WORKER_BAGGAGECLAIM_DRIVER: naive
    networks:
      - appnet

volumes:
  gitea-app:
  gitea-db:

networks:
  appnet:
    external: true
```
### My Pipeline Configuration
```yml
resources:
- name: principal
  type: git
  check_every: 10s
  webhook_token: principal
  source:
    uri: "http://gitea-app:3000/afonsoma9/principal.git"
    branch: develop


jobs:
- name: test
  plan:
    - get: principal
      trigger: true
    - task: run-tests
      config:
        platform: linux
        image_resource:
          type: docker-image
          source:
            repository: circleci/node
            tag: '12.14.1-browsers'
        run:
          path: sh
          args:
                - -c
                - |
                  export NG_CLI_ANALYTICS=ci
                  echo "Node Version: $(node --version)"
                  echo "NPM Version: $(npm --version)"
                  sudo chown -R $USER principal
                  cd principal
                  npm install
                  ls -la
                  whoami
                  npm run test-headless --unsafe-perm
        inputs:
          - name: principal
```

Can anyone help?

**Comment by taylorsilva** [comment]
> What I have: A PR is opened targeting master on my repo / a commit is made to such a PR --> CI picks up on the PR being opened / the commit being made about 10 seconds after it has opened --> Pipeline is started 10-20 seconds or so after PR was opened

@tomgs I think I have a solution for your timing problem here (I hope 🤞). 
tldr: Branch Protection Rules!

Recently in the [concourse-chart] repo we added some checks for PR's. After concourse reported the status of the check back to a PR in `concourse-chart` once, I was then able to set that check as **required** for all future PR's. This results in the 10-20 second gap **never happening**, therefore no one can hit merge without the check happening.

Under the **Settings** tab for the repo, when adding a **branch protection rule** you'll see this area:

<img width="719" alt="Screen Shot 2020-02-16 at 2 38 30 PM" src="https://user-images.githubusercontent.com/4583837/74611473-26099700-50ca-11ea-9dd2-a1071e66427f.png">

Once you make the check required users won't be able to hit merge prematurely:
<img width="661" alt="Screen Shot 2020-02-16 at 2 39 16 PM" src="https://user-images.githubusercontent.com/4583837/74611482-3c175780-50ca-11ea-8cd3-0bd4068e9391.png">

Hope this helps!

[concourse-chart]: https://github.com/concourse/concourse-chart

**Comment by itsdalmo** [comment]
Closing this issue as I believe it has been answered (and also it is more related to Concourse than the resource itself). Feel free to reopen if I'm wrong 🙌
